integration_test/src/test_runner/mod.rs

```diff
@@ -0,0 +1,327 @@
+                    let test_result = match (&test.query, &test.function) {
+                        (Some(_), _) => run_query_test(&test, &connection),
+                        (None, Some(_)) => run_function_test(&test, &connection),
+                        (_, _) => {
+                            return Err(Error::MissingQueryOrFunction(format!(
+                                "{:?}",
+                                test.description
+                            )))
+                        }
+                    };
+                    drop(connection);
+                    return test_result;
+                }
+            }
+        }
+    }
+    Ok(())
+}
+
+fn run_query_test(entry: &TestEntry, conn: &Connection) -> Result<(), Error> {
+    let cursor = conn
+        .execute(entry.query.as_ref().unwrap(), ())
+        .unwrap()
+        .unwrap();
+    validate_result_set(entry, cursor);
+    Ok(())
+}
+
+fn str_or_null(value: &Value) -> *const u8 {
+    if value.is_null() {
+        null_mut()
+    } else {
+        odbc_api::handles::SqlText::new(value.as_str().expect("Unable to cast value as string"))
+            .ptr()
+    }
+}
+
+fn to_i16(value: &Value) -> i16 {
+    value.as_i64().expect("Unable to cast value as i64") as i16
+}
+
+fn check_array_length(array: &Vec<Value>, length: usize) {
+    if array.len() < length {
+        panic!(
+            "not enough values in array expected: {}, actual: {}",
+            length,
+            array.len()
+        )
+    }
+}
+
+fn run_function_test(entry: &TestEntry, conn: &Connection) -> Result<(), Error> {
+    let function = entry.function.as_ref().unwrap();
+    let preallocated = conn.preallocate().unwrap();
+    let statement = preallocated.into_statement();
+    check_array_length(function, 1);
+    let function_name = function[0].as_str().unwrap().to_lowercase();
+    let sql_return = match function_name.as_str() {
+        "sqlgettypeinfo" => {
+            check_array_length(function, 2);
+            unsafe {
+                let data_type: odbc_sys::SqlDataType =
+                    std::mem::transmute(function[1].as_i64().unwrap() as i16);
+                Ok(odbc_sys::SQLGetTypeInfo(statement.as_sys(), data_type))
+            }
+        }
+        "sqltables" => {
+            check_array_length(function, 9);
+            unsafe {
+                Ok(odbc_sys::SQLTables(
+                    statement.as_sys(),
+                    str_or_null(&function[1]),
+                    to_i16(&function[2]),
+                    str_or_null(&function[3]),
+                    to_i16(&function[4]),
+                    str_or_null(&function[5]),
+                    to_i16(&function[6]),
+                    str_or_null(&function[7]),
+                    to_i16(&function[8]),
+                ))
+            }
+        }
+        /*
+        "sqlprimarykeys" => {
+            unsafe {
+                Ok(odbc_sys::SQLPrimaryKeys(
```

Comment:
I'll bring up the missing `SQLPrimaryKeys()` function during standup to see how to handle testing this. 

I added the `SQLForeignKeysW()` function which is in odbc-sys, although the W version in case we want to test it.

Reply:
I created an investigation ticket to address this ([SQL-1015](https://jira.mongodb.org/browse/SQL-1015))